### DISCUSSION TITLE
Troubleshooting Bref Docker Deployment with Octane Handler

### DISCUSSION THREAD

**Discussion by pra-cloud** [brefphp/bref#1780] [Support] [answered]
I want to deploy my code on Docker over Lambda. I obtained this file from the documentation regarding PHP-FPM, but I want to use Octane as a handler. I have made some changes in my Dockerfile, replacing the public/index.php with the Octane handler. After deploying to Lambda, I am encountering this error:

1. **working dockerfile :-** 

FROM bref/php-81-fpm:2
 
COPY --from=bref/extra-redis-php-81:1 /opt /opt
COPY --from=bref/extra-gmp-php-81:1 /opt /opt
 
COPY . /var/task
 
CMD ["public/index.php"]

2. **Not working Dockerfile:-**

FROM bref/php-82-fpm

#COPY --from=bref/extra-redis-php-81:1 /opt /opt
#COPY --from=bref/extra-gmp-php-81:1 /opt /opt

COPY . /var/task

CMD  ["Bref\LaravelBridge\Http\OctaneHandler"]

3. **Error:-    In cloud watch respective lambda function i am getting this error.**

INIT_REPORT Init Duration: 24.44 ms	Phase: init	Status: error	Error Type: Runtime.ExitError

entrypoint requires the handler name to be the first argument

**Comment by mnapoli** [comment]
Hi, you need to switch to `FROM bref/php-82`

**Reply by pra-cloud** [reply]
I have also used base image bref/php-82
But that is also not working.

**Reply by mnapoli** [reply]
Have you installed the laravel bridge?

Also i'm not sure but double check if you need to escape the `\` character in Dockerfile strings (just in case).

**Reply by pra-cloud** [reply]
Yes I have installed laravel-bridge in my project .
and i am using this  CMD ["Bref\LaravelBridge\Http\OctaneHandler"] in a dockerfile , do you have any documentation regarding octane handler with docker.

**Reply by mnapoli** [reply]
No documentation for this scenario explicitly but I'm 99% sure it works, it might be a config issue here. Maybe ARM vs x86?

**Reply by pra-cloud** [reply]
Instead of this handler :- 

CMD["Bref\LaravelBridge\Http\OctaneHandler"]

I have changed this handler in my dockerfile :- 

CMD ["Bref/LaravelBridge/Http/OctaneHandler"]

 but i am getting error for that also:- 
 
 Fatal error: Illuminate\Container\EntryNotFoundException: Bref/LaravelBridge/Http/OctaneHandler in /var/task/vendor/laravel/framework/src/Illuminate/Container/Container.php:748
 
 {
    "message": "Bref/LaravelBridge/Http/OctaneHandler",
    "type": "Illuminate\\Container\\EntryNotFoundException",
    "stackTrace": [
        "#0 /var/task/vendor/bref/laravel-bridge/src/HandlerResolver.php(46): Illuminate\\Container\\Container->get('Bref/LaravelBri...')",
        "#1 /var/task/vendor/bref/bref/src/FunctionRuntime/Main.php(27): Bref\\LaravelBridge\\HandlerResolver->get('Bref/LaravelBri...')",
        "#2 /opt/bref/bootstrap.php(17): Bref\\FunctionRuntime\\Main::run()",
        "#3 {main}"
    ]
}

**Reply by mnapoli** [reply]
This change is incorrect, it should never work and there's no reason to try that.

**Comment by pra-cloud** [comment]
"So, what should I do now? I am unable to run the Octane handler after trying all tweaks in the Dockerfile."
Could you please provide some recommendations or steps to troubleshoot this issue effectively? Any insights or suggestions you can offer would be greatly appreciated.

**Comment by emaadali** [comment]
@mnapoli I'm running into the same issue with this Dockerfile on both ARM and x86:

```dockerfile
FROM bref/php-83:2

COPY . /var/task

CMD ["Bref\LaravelBridge\Http\OctaneHandler"]
```

I also tried escaping the `\` character as you suggested, but no luck:

```dockerfile
FROM bref/php-83:2

COPY . /var/task

CMD ["Bref\\LaravelBridge\\Http\\OctaneHandler"]
```

I even tried referencing the handler directly, knowing it probably wouldn't work:

```dockerfile
FROM bref/php-83:2

COPY . /var/task

CMD ["vendor/bref/laravel-bridge/src/Http/OctaneHandler.php"]
```

---

This might not be directly relevant to the Docker issue, but here's something related to Octane outside of Docker that might provide some insights for debugging:

I've never managed to get the Octane handler to work with Bref v1 syntax when not using Docker. For example:

```yaml
functions:
  web:
    handler: Bref\LaravelBridge\Http\OctaneHandler
    runtime: provided.al2
    layers:
      - ${bref:layer.php-83}
```

This setup throws the error: "Handler `/var/task/Bref\LaravelBridge\Http\OctaneHandler` doesn't exist."

However, this works fine:

```yaml
functions:
  web:
    handler: Bref\LaravelBridge\Http\OctaneHandler
    runtime: php-83
```

Although these should be equivalent, they aren't, and I can't figure out why. I didn't report this before because the Bref v2 syntax works fine.

Do you think these issues could be related?

Any thoughts on what I might check next to troubleshoot the Docker problem?

**Comment by emaadali** [comment]
I figured out a solution. I doubt it is the "correct" solution, but it works.

Create a handler `octane.php`:

```php
<?php

declare(strict_types=1);

use Bref\LaravelBridge\Http\OctaneHandler;

return new OctaneHandler();
```


Use that handler in your Dockerfile:

```Dockerfile
FROM bref/php-83:2

COPY . /var/task

CMD ["octane.php"]
```

**Comment by mnapoli** [comment, marked as answer]
Hey everyone, I finally took some time to try and reproduce.

After debugging, it turns out the CMD arguments must contain valid JSON (https://github.com/moby/moby/issues/23670), so the `\` must be escaped.

So this works:

```dockerfile
CMD ["Bref\\LaravelBridge\\Http\\OctaneHandler"]
```

I've opened https://github.com/brefphp/bref/pull/1821 to document that behavior explicitly.

I'm going to close this as resolved, hopefully that should be the fix for everyone in this thread.

**Reply by emaadali** [reply]
Thanks!
